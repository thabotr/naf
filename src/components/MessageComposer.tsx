/* eslint-disable require-await */
import React from 'react';
import {StyleSheet, View} from 'react-native';
import {Card, TextInput} from 'react-native-paper';

import {useMessageComposer} from '../context/messageEditor';
import {useTheme} from '../context/theme';
import {useLoggedInUser} from '../context/user';
import {useChats} from '../context/chat';
import {HorizontalView} from './Helpers/HorizontalView';
import {useAudioRecorderPlayer} from '../providers/AudioRecorderPlayer';
import {OnlyShow} from './Helpers/OnlyShow';
import {FileManager} from '../services/FileManager';
import {AttachmentsPreview} from './MessageCard/AttachmentsPreview';
import {Remote} from '../services/Remote';
import {AsyncIconButton} from './UserProfile/AsyncIconButton';
import {MutexContextProvider, useMutex} from '../providers/MutexProvider';

const EditorActions = () => {
  const {userProfile} = useLoggedInUser();
  const {addChatMessages, activeChat} = useChats();

  const {
    composeMsg,
    saveComposeMsg,
    saveComposeState,
    composeState,
    addAttachments,
  } = useMessageComposer();
  const {startRecorder} = useAudioRecorderPlayer();

  const interlocutor = activeChat()?.user;

  const openCamInMode = (mode: 'video' | 'photo') => {
    FileManager.getCameraMedia(mode)
      .then(vidOrPic => {
        vidOrPic &&
          saveComposeMsg(msg => {
            if (msg) {
              return {
                ...msg,
                files: [...msg.files, vidOrPic],
              };
            } else {
              return {
                from: userProfile.handle,
                to: interlocutor.handle,
                id: new Date().getTime(),
                files: [vidOrPic],
                voiceRecordings: [],
              };
            }
          });
      })
      .catch(e => console.warn('camera error', e));
  };

  const saveDraft = async () => {
    composeMsg &&
      addChatMessages([
        {
          ...composeMsg,
          id: new Date().getTime(),
          from: userProfile.handle ?? '',
          to: activeChat()?.user.handle ?? '',
          draft: true,
        },
      ]);
    saveComposeMsg(_ => undefined);
  };

  const sendMessage = async () => {
    if (composeMsg) {
      const msgRes = await Remote.sendMessage(
        userProfile.token,
        userProfile.handle,
        {
          ...composeMsg,
          from: userProfile.handle,
          to: interlocutor.handle,
        },
      );
      if (msgRes) {
        saveComposeMsg(_ => undefined);
        addChatMessages([msgRes]);
      } else {
        throw new Error('failed to send message draft');
      }
    }
  };

  return (
    // eslint-disable-next-line react-native/no-inline-styles
    <HorizontalView style={{justifyContent: 'space-between'}}>
      <HorizontalView>
        <AsyncIconButton
          icon="content-save-edit"
          disabled={!composeMsg?.text && !composeMsg?.files.length}
          onPress={saveDraft}
        />
        <AsyncIconButton
          icon="delete"
          onPress={async () => saveComposeMsg(_ => undefined)}
        />
      </HorizontalView>

      <HorizontalView>
        <AsyncIconButton
          icon="emoticon-excited-outline"
          onPress={async () => {}}
        />
        <AsyncIconButton
          icon="microphone"
          onPress={async () => startRecorder()}
        />
        <AsyncIconButton
          icon="attachment"
          onPress={async () => addAttachments()}
        />
        <AsyncIconButton
          icon="camera"
          onPress={async () => openCamInMode('photo')}
        />
        <AsyncIconButton
          icon="video"
          onPress={async () => openCamInMode('video')}
        />
        <AsyncIconButton
          icon={composeState.inputTextEnabled ? 'pencil-minus' : 'pencil-plus'}
          onPress={async () => {
            saveComposeMsg(msg => {
              if (msg) {
                return {
                  ...msg,
                  text: undefined,
                };
              }
            });
            saveComposeState(cs => {
              return {
                ...cs,
                inputTextEnabled: !cs.inputTextEnabled,
              };
            });
          }}
        />
      </HorizontalView>

      <AsyncIconButton
        icon="send"
        onPress={sendMessage}
        disabled={
          !(
            composeMsg?.text ||
            composeMsg?.files.length ||
            composeMsg?.voiceRecordings.length
          )
        }
      />
    </HorizontalView>
  );
};

export const MessageEditorCard = () => {
  const {theme} = useTheme();
  const {activeChat} = useChats();

  const {composeMsg, saveComposeMsg, composeState} = useMessageComposer();

  const interlocutor = activeChat()?.user;

  if (!interlocutor) {
    return <></>;
  }

  const styles = StyleSheet.create({
    editorContainer: {
      backgroundColor: theme.color.userPrimary,
      margin: 2,
      paddingBottom: 5,
    },
    editorBody: {padding: 10},
    fullWidth: {width: '100%'},
  });

  const saveMessageTxt = (text: string) => {
    saveComposeMsg(msg => {
      if (msg) {
        return {
          ...msg,
          text: text,
        };
      }
    });
  };

  const EditorTextInput = () => {
    const {slots} = useMutex();
    return (
      <OnlyShow If={!!composeMsg && composeState.inputTextEnabled}>
        <TextInput
          defaultValue={composeMsg?.text}
          onEndEditing={e => saveMessageTxt(e.nativeEvent.text)}
          disabled={!slots}
          multiline
          numberOfLines={6}
          style={styles.fullWidth}
          label="message body"
        />
      </OnlyShow>
    );
  };

  return (
    <OnlyShow If={!!composeMsg}>
      <MutexContextProvider>
        <Card style={styles.editorContainer}>
          <View style={styles.editorBody}>
            <EditorActions />
            <EditorTextInput />
            {composeMsg && <AttachmentsPreview msg={composeMsg} composing />}
          </View>
        </Card>
      </MutexContextProvider>
    </OnlyShow>
  );
};
