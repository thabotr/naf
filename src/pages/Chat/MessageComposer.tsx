import React from 'react';
import {Surface, TextInput} from 'react-native-paper';
import globalStyles from '../../shared/styles';
export default function () {
  return (
    <Surface accessibilityLabel="message composer" style={globalStyles.card}>
      <TextInput label={'message text'} />
    </Surface>
  );
}
