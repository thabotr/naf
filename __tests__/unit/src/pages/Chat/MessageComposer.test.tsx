import {fireEvent, render, screen} from '@testing-library/react-native';
import React from 'react';
import MessageComposer from '../../../../../src/pages/Chat/MessageComposer';
import themed from '../../../utils/themed';

describe('Chat page', () => {
  describe('MessageComposer', () => {
    const inititalText = 'hello';
    test(
      'when the initial message text is falsy then the text field should be ' +
        'empty and the send button should be disabled',
      () => {
        render(themed(<MessageComposer />));
        const textInputField = screen.getByLabelText('message text input');
        expect(textInputField.props.defaultValue).toBeFalsy();
        const sendButton = screen.getByLabelText('send message');
        expect(sendButton.props.accessibilityState.disabled).toBeTruthy();
      },
    );
    test(
      'when the initial message text is given then the text field should contain ' +
        'that text and the send button should be enabled',
      () => {
        render(
          themed(<MessageComposer initialMessage={{text: inititalText}} />),
        );
        const textInputField = screen.getByLabelText('message text input');
        expect(textInputField.props.defaultValue).toBe(inititalText);
        const sendButton = screen.getByLabelText('send message');
        expect(sendButton.props.accessibilityState.disabled).toBeFalsy();
      },
    );
    it('disables the send button when the text input field is cleared', () => {
      render(themed(<MessageComposer initialMessage={{text: inititalText}} />));
      const textInputField = screen.getByLabelText('message text input');
      fireEvent.changeText(textInputField, '');
      const sendButton = screen.getByLabelText('send message');
      expect(sendButton.props.accessibilityState.disabled).toBeTruthy();
    });
    it('enables the send button when the text input field is populated', () => {
      render(themed(<MessageComposer />));
      const textInputField = screen.getByLabelText('message text input');
      const someText = 'a';
      fireEvent.changeText(textInputField, someText);
      const sendButton = screen.getByLabelText('send message');
      expect(sendButton.props.accessibilityState.disabled).toBeFalsy();
    });
    it(
      "should call the 'onSendMessage' prop with the composed message when " +
        'the send message button is clicked',
      () => {
        const mockOnSendMessage = jest.fn().mockName('mockOnSendMessage');
        const composedMessage = {text: inititalText};
        render(
          themed(
            <MessageComposer
              initialMessage={composedMessage}
              onSendMessage={mockOnSendMessage}
            />,
          ),
        );
        const sendButton = screen.getByLabelText('send message');
        fireEvent.press(sendButton);
        expect(mockOnSendMessage).toBeCalledTimes(1);
        expect(mockOnSendMessage).toBeCalledWith(composedMessage);
      },
    );
    it(
      "should call the 'onDiscardMessage' prop with the composed message when " +
        'the discard message button is clicked',
      () => {
        const mockOnDiscardMessage = jest.fn().mockName('mockOnDiscardMessage');
        const composedMessage = {text: inititalText};
        render(
          themed(
            <MessageComposer
              initialMessage={composedMessage}
              onDiscardMessage={mockOnDiscardMessage}
            />,
          ),
        );
        const discardButton = screen.getByLabelText('discard message');
        fireEvent.press(discardButton);
        expect(mockOnDiscardMessage).toBeCalledTimes(1);
        expect(mockOnDiscardMessage).toBeCalledWith(composedMessage);
      },
    );
  });
});
