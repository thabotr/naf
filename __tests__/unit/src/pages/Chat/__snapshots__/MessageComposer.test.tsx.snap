// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat MessageComposer renders correctly 1`] = `
<View
  accessibilityLabel="message composer"
  collapsable={false}
  style={
    {
      "backgroundColor": "#ffffff",
      "elevation": 3,
      "margin": 3,
      "shadowColor": "#000000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.24,
      "shadowRadius": 3,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "rgb(231, 231, 231)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        {},
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(0, 0, 0, 0.26)",
          "bottom": 0,
          "height": 2,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "scaleY": 0.5,
            },
          ],
          "zIndex": 1,
        }
      }
    />
    <View
      style={
        [
          {
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          {
            "minHeight": 64,
          },
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="none"
        style={
          {
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "zIndex": 3,
          }
        }
      >
        <Text
          collapsable={false}
          maxFontSizeMultiplier={1.5}
          numberOfLines={1}
          onLayout={[Function]}
          style={
            {
              "color": "#6200ee",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "left": 0,
              "opacity": 0,
              "paddingLeft": 12,
              "paddingRight": 12,
              "position": "absolute",
              "textAlign": "left",
              "top": 34,
              "transform": [
                {
                  "translateX": 0,
                },
                {
                  "translateY": 0,
                },
                {
                  "scale": 1,
                },
              ],
              "writingDirection": "ltr",
            }
          }
        >
          message text
        </Text>
        <Text
          collapsable={false}
          maxFontSizeMultiplier={1.5}
          numberOfLines={1}
          style={
            {
              "color": "rgba(0, 0, 0, 0.54)",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "left": 0,
              "opacity": 0,
              "paddingLeft": 12,
              "paddingRight": 12,
              "position": "absolute",
              "textAlign": "left",
              "top": 34,
              "transform": [
                {
                  "translateX": 0,
                },
                {
                  "translateY": 0,
                },
                {
                  "scale": 1,
                },
              ],
              "writingDirection": "ltr",
            }
          }
        >
          message text
        </Text>
      </View>
      <TextInput
        editable={true}
        maxFontSizeMultiplier={1.5}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder=" "
        placeholderTextColor="rgba(0, 0, 0, 0.54)"
        selectionColor="#6200ee"
        style={
          [
            {
              "flexGrow": 1,
              "margin": 0,
            },
            {
              "paddingLeft": 12,
              "paddingRight": 12,
            },
            {
              "height": 64,
            },
            {
              "paddingBottom": 4,
              "paddingTop": 24,
            },
            {
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 16,
              "fontVariant": undefined,
              "fontWeight": undefined,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            [
              {},
            ],
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
</View>
`;
